Propagate the changes in some transactions columns through the rest of

lomamech opened the pull request:
Propagate the changes in some transactions columns through the rest of  transaction
- Implentation of function propagate
- Improve date in data in bhima_test Database

closes #1618

lomamech committed:
Propagate the changes in some transactions columns through the rest of
transaction
- Implentation of function propagate
- Improve date in data in bhima_test Database

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -277,6 +277,33 @@ function JournalController(Journal, Sorting, Grouping,
   function onRegisterApi(gridApi) {
     vm.gridApi = gridApi;
+    
+    vm.gridApi.edit.on.afterCellEdit($scope, function(rowEntity, colDef, newValue, oldValue){
```
:+1:

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -512,6 +539,9 @@ function JournalController(Journal, Sorting, Grouping,
   function editTransactionType(row) {
```
We should eventually make this a `ui-grid` dropdown editor - it would remove some code, and you wouldn't have to call `propogate` (because it would be caught by the `afterCellEdit` method you created).

I'll make an issue about it - no need to address it in the PR.

jniles replied:
Issue created: https://github.com/IMA-WorldHealth/bhima-2.X/issues/1643

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -277,6 +277,33 @@ function JournalController(Journal, Sorting, Grouping,
+    var propagateColumn = ['trans_date', 'entity_uuid', 'origin_id'];
+    
+    //Check if the column updated must be propragated in all transaction
+    var checkPropagate = (propagateColumn.indexOf(column));
```
A clearer way to do write this is to do the ` value === -1` check here.  Like this:
```js
var hasSharedProperty = propagateColumn.indexOf(column) !== -1;

if (hasSharedProperty) {
  // propagate changes to all rows!
}
```

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -277,6 +277,33 @@ function JournalController(Journal, Sorting, Grouping,
+
+    if(checkPropagate !== -1){
+      // For the old row
+      vm.transactions._entity.data.data.forEach(function (row) {
```
Lines 296-299 is the same code as lines 302-305.  Let's make it into the same funtion:

```js
function updateSharedPropertyOnRow(row) {
  transactions.editCell(row, column, value, row[column]);
  row[column] = (column === 'trans_date') ? new Date(value) : value;
}

// pass updates on to both the original rows and the new (pending) rows
vm.transactions._entity.data.data.forEach(updateSharedPropertyOnRow);
vm.transactions._entity.newRows.data.forEach(updateSharedPropertyOnRow);
```

jniles replied:
~~If you move this function outside of the `propagate` function, it will reduce the number of memory allocations that are performed.   V8 will not garbage collect it as often.~~
```js
function updateSharedPropertyOnRow(row) {
 // function body
}

function propagate(column, value) {
  // function body
}
```

*Edit*: this won't work since you depend on variables local to `propagate()`.  Ignore this comment.

jniles reviewed (approved):
@lomamech this PR is exactly correct.

I've made some comments to improve the clarity of the code (and a tiny bump to memory performance).  Let me know what you think of these suggestions - the code could be merged without them if you think they are not improvements.

lomamech committed:
Improve the code
- using a clearer way to use shared property
- Adding function updateSharedProperty for reduice the memory allocations

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -285,24 +285,23 @@ function JournalController(Journal, Sorting, Grouping,
-        transactions.editCell(row, column, value, row[column]);
-        row[column] = (column === 'trans_date') ? new Date(value) : value;
-      });
+    if (hasSharedProperty) {
```
Beautiful!  This is a better proposal than my suggestion. :+1:
